### ISSUE TITLE
Fix the weight and sigma (and spectrum) fudgery

### ISSUE THREAD

**Issue by bennahugo** [tart-telescope/tart2ms#18]
This needs to be fixed for the sake of imaging properly

I know MeerKAT uses something called Wishart noise statistics based on the autocorrelations. @tmolteno do we get correct autos out of the telescope? @landmanbester can you enlighten me  about the correct derivation of these. I need just the auto powers and the correlator counts  that goes into each sample right? In first order if we assume things are not being flagged by the X engine within integrations can we just scale by a normalized auto power?

**Comment by tmolteno** [comment]
We certainly don't have good estimates of sigma for these. There was a wonderful talk by one of the SARAO engineers on this a couple of years back. The problem is that some antennas are very close together and so noise is also correlated between them. Urgh. I am hoping to incorporate this into the array optimization code at some point.

**Comment by bennahugo** [comment]
I will see if I can image either way. For the school it might suffice for now, but I think this ticket should remain open

**Comment by landmanbester** [comment]
Yep, Ludwig gave a good talk about this. We should really get him to write it down. There is a recursive formula to write done the moments of a proper/circular complex Wishart distribution. To quote Ludwig

>The actual weights are inverse variances of the corresponding visibilities [1],
allowing proper statistical combination e.g. when doing weighted averaging.

>The variance of visibility v_ij is v_ii * v_jj / K, in other words the product of the
corresponding autocorrelations divided by the number of samples that went into
calculating the visibility. This is based on Wishart statistics. The weight is therefore
K / (v_ii * v_jj). This is accessed in katdal as the `weights` property on the dataset
(and is written to the WEIGHT_SPECTRUM column when converting to MS).

>The number K is generally t_dump * f_sample / 2 / n_chans, i.e. the number of
F-engine samples in the SDP dump. K can be lower due to excision, which typically
happens when captures stop and start, between tracks and slews, and when some
CBF dumps making up an SDP dump are flagged for RFI in ingest. The excision
fraction is accessed in katdal as the `excision` property on the dataset object.

>Some implementation details: We used to calculate the total weight in ingest on the
GPUs, but now the division by autocorrelations happens in katdal. Ingest only produces K.
The original ingest scheme used lossy compression to store the bulk of the weights
as single bytes, by supplementing them with floating-point scale factors per channel
(called "weights_channel"). This scheme persists, even though we are now effectively
only storing the number of CBF dumps per SDP dump in the archive. 

So, if you don't have *K*, you can approximate it as *t_dump * f_sample / 2 / n_chans*

**Comment by bennahugo** [comment]
Presuming @tmolteno does not do anything super duper fancy like a rank type filter (as we have with MeerKAT ingest -- think it is  a MAD based thing if I recall) N should be constant on both the autos. WIll it be enough to approximate as  v_ii * v_jj presuming that N is fixed over the entire span of the observation?

Alternatively what I have thought (and this is much more computationally intensive) is to approximate noise by subtraction of the even and odd visibilities. on short intervals this completely removes Tsky from the equation and leaves you with Tadc + Tspill + Tatm.

Thoughts @landmanbester

**Comment by landmanbester** [comment]
I suspect the product of the autos will serve as a good proxy for the weights (they should work well for imaging and maximum likelihood calibration). What we can do, as a consistency check, is compare the empirical estimates derived from the difference between odd and even time stamps, which will be far noisier, to the weights derived from the autos. The ratios of the medians for a baseline should be approximately N. I would only start worrying about this if we find that these ratios differ significantly between different baselines. We have some more sophisticated machinery we could throw at the problem if we find that to be the case...

**Comment by bennahugo** [comment]
Yea I agree the even/odd differences will be noisy - no correlated power
remains and may need to be a mean (or short term mean) that broadcasts into
the per row terms. I like the autos better presuming they are properly
formed.

Something I don't understand though is how this yields a good estimator for
MeerKAT at low elevations where Tspill introduces correlated reflections
across the band on autos and completely decorrelated Tspill on the crosses.
If we use the autos with this Tspill passband ripping included it cannot be
a good estimator at low elevations  (maybe only on the shortest of
spacings).

This is not the case for TART ofc. but it is something that has been
bugging me for a while.

**Comment by landmanbester** [comment]
I think you are trying to think in terms of the physics whereas the derivation really comes from statistical considerations which all follow from the fact that visibility matrices are samples from a (complex proper) Wishart distribution by construction. The variance of a visibility can be computed as the second central moment of one of the elements of the Wishart matrix. This is not perfect because we only have a single sample from the governing Wishard distribution but it should be a good approximation. One thing to be wary of, and maybe this is where your concerns are justified, is that the values derived in this way do not only relate to the astrophysical signal but to the combined signal from all sources. You still need to flag out the bad data, it would just be too magical otherwise

**Comment by tmolteno** [comment]
We can compute all these fairly easily as we have access to the raw samples during the calibration process (which is run every 2 hours).

**Comment by bennahugo** [comment]
Thanks Tim

Ok so as discussed these are not in the provided X engine visibilities you
can pull, but they can be formed from raw filterbank data products.

My suggestion is as follows (it is going to cost some storage though). We
have to store only SIGMA_SPECTRUM. I will pull the formulas for you to use
when I get to the office.  This must be treated as an optional column by
all export software. If it does not exist we replace with perhaps short
time averages of the even / odd differences as an approximation. This is
needed to keep conformance with older databases.

Thanks @landmanbester for your inputs - yes I was only worried because the
input signals are not actually the same on a highly directional system.

I will have a look at if  Rhodes data can be salvaged today with
appropriate downweighting.

**Comment by tmolteno** [comment]
You can download data to play with:
```    tart_download_data --raw --n 1 --pw <password>```

This will download the raw data files. Then the example code is in https://github.com/tmolteno/TART/blob/next/doc/getting_started/autocorrelation.py, and run it with:
```    python3 autocorrelation.py --file data_2022-10-26_16_00_24.852782.hdf```

Here are the first 4 antennas autocorrelations from the NZ TART:
![autocorrelation](https://user-images.githubusercontent.com/85630/198393664-5dee095d-893f-4a8e-bd1e-a0fab0b8755c.png)
